Docs: Checkout (#5606)
vuestorefront/vue-storefront: packages/core/docs/guide/checkout.md

# Checkout
Checkout is a process of providing information about shipping and billing details, and picking certain shipping and payment method. Then with all the data, we are able to place an order and pay for it.

> ## Shipping

Reviewer: as @NataliaTepluhina was suggesting let's have more descriptive subtitles (also for SEO)

> To access shipping details, we can use property of `useShipping` called `shipping`.

Reviewer: maybe its worth explaining very briefly what are shipping details

```js{8,16}
import { useShipping } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      load,
      shipping
    } = useShipping();

    onSSR(async () => {
      await load();
    });

    return {
      shipping
    };
  }
}
```

> `shipping` property will return `null` if the address is not saved on the server or just has not been loaded. In order to save shipping details, you could use `save` method:

Reviewer: ```suggestion
`shipping` property will return `null` if the address is not saved on the server or just has not been loaded. In order to save shipping details, you can use `save` method:
```

also it's worth explaining what `save` means a bit further (eg it will be persisted in a composable state / available each time you call `load` or similar )

```vue{2,15,24}
<template>
  <form @submit.prevent="save({ shippingDetails: shippingForm })">
    <!-- form fields -->
    <button type="submit" :disabled="loading">Submit</button>
  </form>
</template>

<script>
import { useShipping } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      save,
      loading,
      shipping
    } = useShipping();

    const shippingForm = ref(/* object for shipping details */);

    return {
      shippingForm,
      save
    };
  }
}
</script>
```

> ## Shipping providers

Reviewer: The below paragraph will probably be unable to understand by anyone who doesn't already know what shipping provider component is. The explanation is missing essential information like why we need this component, what is a shipping provider component, what are responsibilities of such component etc.

Reviewer: also what is a proper place, please avoid such constructions in the docs - be specific

We delegate whole logic of selecting shipping method to the dedicated component called **VsfShippingProvider.vue**. Each shipping extension shares one. All we have to do is import and put it in a proper place.
```vue
<VsfShippingProvider
  @submit="$router.push('/checkout/billing')"
/>
```

> VsfShippingProvider emits `submit` event when it is ready to move to the next step.

Reviewer: ```suggestion
VsfShippingProvider emits `submit` event when it is ready to move to the next step.
```
how can component be ready to move? :D

> Component shares 5 hooks with async support:

Reviewer: ```suggestion
You can pass asynchronous functions as `VsfShippingProvider` props to hook into different events within it's lifecycle and override initial function parameters or react to specific events like method selection
```

- **beforeLoad** - called before loading shipping methods
- **afterLoad** - called after loading shipping methods
- **onSelected** - called after selecting shipping method
- **onSelectedDetailsChanged** - called after modifying currently picked shipping method, e.g. selecting parcel locker on the map
- **onError** - called when some operation throws an error

> We can inject in the hook via props:

Reviewer: we can just have a code with comments + with functions in a script section (or their types below) instead of a list AND code. 

You will avoid repetition by doing sth like this:

```
<VsfShippingProvider
  <!-- called before loading shipping methods -->
  :beforeLoad="(config) => config"
  <!-- ... -->
  :afterLoad="..."
  :onSelected="..."
  :onSelectedDetailsChanged="..."
  :onError="..."
/>

```

Author: It isn't a proper markup, so the colors under the line with the comment would be not proper too.
![image](https://user-images.githubusercontent.com/30155292/109982002-c34d6280-7d01-11eb-842b-96af449af204.png)

```vue
<VsfShippingProvider
  :beforeLoad="beforeLoad"
  :afterLoad="afterLoad"
  :onSelected="onSelected"
  :onSelectedDetailsChanged="onSelectedDetailsChanged"
  :onError="onError"
/>
```

> ## Billing

Reviewer: all comments from above apply here

```js{8,16}
import { useShipping } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      load,
      shipping
    } = useShipping();

    onSSR(async () => {
      await load();
    });

    return {
      shipping
    };
  }
}
```
`shipping` property will return `null` if the address is not saved on the server or has not been loaded.   

In order to save shipping details on the server, you can use `save` method:
```vue{2,15,24}
<template>
  <form @submit.prevent="save({ shippingDetails: shippingForm })">
    <!-- form fields -->
    <button type="submit" :disabled="loading">Submit</button>
  </form>
</template>

<script>
import { useShipping } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      save,
      loading,
      shipping
    } = useShipping();

    const shippingForm = ref(/* object for shipping details */);

    return {
      shippingForm,
      save
    };
  }
}
</script>
```

Shipping details stored on the server with `save` method will be possible to load after a refresh with `load` method.

## Shipping providers

> Shipping provider is an aggregator that provides us one or more shipping methods.    

Reviewer 2: please mention that's integration, not only the "aggregator", always one provider components mean one 3rd party provider of either payments or shippings

- Loading and showing available shipping methods
- Loading currently picked shipping method
- Showing UI
- Picking and configuring shipping method

All we have to do is import and put it in `pages/Shipping.vue` as a [second part of the Shipping step](https://github.com/vuestorefront/vue-storefront/blob/next/packages/commercetools/theme/pages/Checkout/Shipping.vue#L197)
```vue
<VsfShippingProvider
  @submit="$router.push('/checkout/billing')"
/>
```

`VsfShippingProvider` emits `submit` event when shipping method is properly picked, configured and user clicks `Continue to billing` button.

### Hooks
You can pass asynchronous functions as `VsfShippingProvider` props to hook into different events within it's lifecycle and override initial function parameters or react to specific events like method selection
- **beforeLoad** - called before loading shipping methods
- **afterLoad** - called after loading shipping methods
- **onSelected** - called after selecting shipping method
- **onSelectedDetailsChanged** - called after modifying currently picked shipping method, e.g. selecting parcel locker on the map

> - **onError** - called when some operation throws an error

Reviewer 2: mention signature: what arguments I have and what values they are returning

```js{8,16}
import { useBilling } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      load,
      billing
    } = useBilling();

    onSSR(async () => {
      await load();
    });

    return {
      billing
    };
  }
}
```
`billing` property will return `null` if the address is not saved on the server or has not been loaded.   

In order to save billing details on the server, you could use `save` method:
```vue{2,15,24}
<template>
  <form @submit.prevent="save({ billingDetails: billingForm })">
    <!-- form fields -->
    <button type="submit" :disabled="loading">Submit</button>
  </form>
</template>

<script>
import { useBilling } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      save,
      loading,
      billing
    } = useBilling();

    const billingForm = ref(/* object for billing details */);

    return {
      billingForm,
      save
    };
  }
}
</script>
```
Billing details stored on the server with `save` method will be possible to load after a refresh with `load` method.

## Payment providers

> Payment provider is an aggregator that provides us one or more payment methods.    

Reviewer 2: same here

import { useOrder } from '{INTEGRATION}';

export default {
  setup () {
    // ...
    const { make } = useOrder();
    const { pay } = usePayment();

    const placeAnOrder = () => {
      await make();
      await pay();
    }
  }
}
```

Having a composable makes it so easy for integrator to share data between Vue's component and `pay` method via [sharedRef](https://github.com/vuestorefront/vue-storefront/blob/next/packages/core/core/src/utils/shared/index.ts).

`VsfPaymentProvider` emits `status` event with boolean in the payload which tells whether everything is ready for call `pay` or not. Use this information to enable/disable `Place order` button.

### Hooks 
You can pass asynchronous functions as `VsfPaymentProvider` props to hook into different events within it's lifecycle and override initial function parameters or react to specific events like method selection
- **beforeLoad** - called before loading payment methods
- **afterLoad** - called after loading payment methods
- **onSelected** - called after selecting payment method
- **onSelectedDetailsChanged** - called after modifying currently picked payment method, e.g. changing credit card's details

> - **onError** - called when some operation throws an error

Reviewer 2: and here

# What's checkout?
Checkout is a process of providing information about shipping and billing details, and picking certain shipping and payment method. Then with all the data, we are able to place an order and pay for it.

## Shipping details
Shipping details are information about the receiver and her/his address. Based on that, shop will know how to address a package.

To access shipping details, we can use property returned by `useShipping` called `shipping`.
```js{8,16}
import { useShipping } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      load,
      shipping
    } = useShipping();

    onSSR(async () => {
      await load();
    });

    return {
      shipping
    };
  }
}
```

> `shipping` property will return `null` if the address is not saved on the server or has not been loaded.   

Author: `or` fits pretty well there in my opinion.

I could rephrase it to smth like that at first (`or` still present):
`shipping` property returns `null` if the `load` function was not invoked or the `save` function was not invoked before

# Checkout
Checkout is a process of providing information about shipping and billing details, and picking certain shipping and payment method. Then with all the data, we are able to place an order and pay for it.

## Collecting and saving shipping details
Shipping details are information about the receiver and her/his address. Based on that, we know where to send the order.

We can load shipping details by invoking `load` method. To access them, use property returned by `useShipping` called `shipping`.
```js{8,16}
import { useShipping } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      load,
      shipping
    } = useShipping();

    onSSR(async () => {
      await load();
    });

    return {
      shipping
    };
  }
}
```

> `shipping` property returns `null` if the `load` function was not invoked or nothing is saved.    

Author: We cannot assume it is always saved on the server. Also @filrak wanted to make it clear that we need to call `load` (in his CR above)

> You can use `save` method, in order to save shipping details.

Author: Fair enough

To access billing details, we can use property of `useBilling` called `billing`.
```js{8,16}
import { useBilling } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      load,
      billing
    } = useBilling();

    onSSR(async () => {
      await load();
    });

    return {
      billing
    };
  }
}
```
`billing` property will return `null` if the address is not saved on the server or has not been loaded.   

In order to save billing details on the server, you could use `save` method:
```vue{2,15,24}
<template>
  <form @submit.prevent="save({ billingDetails: billingForm })">
    <!-- form fields -->
    <button type="submit" :disabled="loading">Submit</button>
  </form>
</template>

<script>
import { useBilling } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      save,
      loading,
      billing
    } = useBilling();

    const billingForm = ref(/* object for billing details */);

    return {
      billingForm,
      save
    };
  }
}
</script>
```

> Billing details stored on the server with `save` method will be possible to load after a refresh with `load` method.

Reviewer 3: ```suggestion

```

## Making an order
After providing every crucial information by the user, we are ready to *make an order*. To do that, we have to call a `make` method from the `useMakeOrder` composable.
```js
import { useMakeOrder } from '{INTEGRATION}';

export default {
  setup () {
    const { make } = useMakeOrder();

    return {
      make
    }
  }
}
```

> It creates an order but we need to perform additional actions:
> - redirect to thank you page
> - clear a cart

Reviewer 3: ```suggestion
When the order is created, we can redirect the user to the page thanking them for making an order and refresh the cart.
```

`It creates an order but we need to perform additional actions:` sounds like there are some other steps required to finalize the order.

```js
import { useMakeOrder, useCart } from '{INTEGRATION}';

export default {
  setup (_, context) {
    const { make, order } = useMakeOrder();
    const { setCart } = useCart();

    const processOrder = async () => {
      await make();
      context.root.$router.push(`/checkout/thank-you?order=${order.value.id}`);
      setCart(null);
    }
  }
}
```

## Payment providers

> Payment provider is an aggregator that provides us one or more payment methods. It is also an integration, one provider component always means one 3rd party provider of payments.       

Reviewer 3: ```suggestion
A payment provider is a component that provides one or more payment methods. One such component integrates one 3rd party provider of payments.
```

> To give you the best developer experience, we delegate whole logic of selecting and configuring payment method to the dedicated component called **VsfPaymentProvider.vue**. It takes care of:

Reviewer 3: ```suggestion
For the best developer experience, we delegate the whole logic of selecting and configuring payment methods to the dedicated component named **VsfPaymentProvider.vue**. It takes care of:
```

- Loading and showing available payment methods

> - Picking and configuring payment method

Reviewer 3: List punctuation

> The first thing we have to do is import and put the component inside `pages/Payment.vue` as a [last part of the Checkout](https://github.com/vuestorefront/vue-storefront/blob/next/packages/commercetools/theme/pages/Checkout/Payment.vue#L108):

Reviewer 3: ```suggestion
All we have to do is to use it in the `pages/Checkout/Payment.vue` as a [last part of the Checkout](https://github.com/vuestorefront/vue-storefront/blob/next/packages/commercetools/theme/pages/Checkout/Payment.vue#L108):
```

```vue
<template>
  <!-- ... -->
  <VsfPaymentProvider />
  <!-- ... -->
</template>

<script>
import VsfPaymentProvider from '{PAYMENT_PROVIDER_INTEGRATION}/src/VsfPaymentProvider';

export default {
  components: {
    VsfPaymentProvider
  }
}
</script>
```

> Then we have to support making a payment - each package with a payment provider might force us to use a bit different approach. We found that there will be 2 most common ones:

Reviewer 3: ```suggestion
The next step is making a payment. Each package with a payment provider might use a slightly different approach, but below we described are two most common.
```

We can load billing details by calling the `load` method in `useBilling` composable and use the `billing` property to access it.
```js{8,16}
import { useBilling } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      load,
      billing
    } = useBilling();

    onSSR(async () => {
      await load();
    });

    return {
      billing
    };
  }
}
```
`billing` property returns `null` if the `load` function was not invoked or nothing is saved.   

We can use `save` method to save billing details.
```vue{2,15,24}
<template>
  <form @submit.prevent="save({ billingDetails: billingForm })">
    <!-- form fields -->
    <button type="submit" :disabled="loading">{{ $t('Submit') }}</button>
  </form>
</template>

<script>
import { useBilling } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      save,
      loading,
      billing
    } = useBilling();

    const billingForm = ref(/* object for billing details */);

    return {
      billingForm,
      save
    };
  }
}
</script>
```

> ## Making an order

Reviewer: I know in many cases the order is opposite but maybe we should have this section at the end?

Author: I am not sure about that. In PaymentProvider part, I mention about doing payment after order and using `useMakeOrder` which requires to put Order part before.

- **beforeSelect** `(shippingMethod => shippingMethod)` - Called before selecting payment method.
- **afterSelect** `(selectedShippingMethod => void)` - Called after selecting payment method.
- **beforePay** `(paymentDetails => paymentDetails)` - Called before pay.
- **afterPay** `(paymentResponse => void)` - Called after pay.
- **beforeSelectedDetailsChange** `(details => details)` - Called before modifying currently picked payment method, e.g. changing credit card's details.
- **afterSelectedDetailsChange** `(details => void)` - Called after modifying currently picked payment method, e.g. changing credit card's details.
- **onError** `(({ action, error }) => void)` - Called when some operation throws an error.

```vue
<VsfPaymentProvider
  :beforeLoad="beforeLoad"
  :afterLoad="afterLoad"
  :beforeSelect="beforeSelect"
  :afterSelect="afterSelect"
  :beforePay="beforePay"
  :afterPay="afterPay"
  :beforeSelectedDetailsChange="beforeSelectedDetailsChange"
  :afterSelectedDetailsChange="afterSelectedDetailsChange"
  :onError="onError"
/>
```

### Why some integrations have a mocked VsfPaymentProvider?

> There are eCommerce backends which do not provide any payment methods out-of-the-box, e.g. commercetools. For these, we provide mocked component to let user go through the whole checkout. We are using external providers with dedicated VsfPaymentProvider for them. One example is a [Checkout.com integration](https://github.com/vuestorefront/checkout-com).

Reviewer: ```suggestion
There are eCommerce backends which do not provide any payment methods out-of-the-box, e.g. commercetools. For these, we provide mock component that has exactly same interface as a real integration so you can easily swap it with a payment integration of your choice. You can find available payment integrations [here](/v2/integrations/)
```

```js{8,16}
import { useBilling } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      load,
      billing
    } = useBilling();

    onSSR(async () => {
      await load();
    });

    return {
      billing
    };
  }
}
```
`billing` property will return `null` if the address is not saved on the server or has not been loaded.   

In order to save billing details on the server, you could use `save` method:
```vue{2,15,24}
<template>
  <form @submit.prevent="save({ billingDetails: billingForm })">
    <!-- form fields -->
    <button type="submit" :disabled="loading">Submit</button>
  </form>
</template>

<script>
import { useBilling } from '{INTEGRATION}';
import { onSSR } from '@vue-storefront/core';

export default {
  setup () {
    const {
      save,
      loading,
      billing
    } = useBilling();

    const billingForm = ref(/* object for billing details */);

    return {
      billingForm,
      save
    };
  }
}
</script>
```
Billing details stored on the server with `save` method will be possible to load after a refresh with `load` method.

## Order

> After providing every crucial information by the user, we are ready to *make an order*. To do that, we have to call a `make` method from the `useMakeOrder` composable.

Reviewer 3: ```suggestion
Once the user provides all required information, we can *make an order* by calling the `make` method in `useMakeOrder` composable.
```

```

### Extending `VsfPaymentProvider` and reacting to it's events
We can pass asynchronous functions to the `VsfPaymentProvider` component to hook into different events within its lifecycle, override initial function parameters or react to specific events like method selection. We will call these methods "hooks".

Because every payment provider is different, not all of them are present in every integration. Always refer to the documentation of a specific provider to learn which hooks are available.

> - **beforeLoad** `(config => config)` - Called before loading payment methods/

Reviewer 3: ```suggestion
- **beforeLoad** `(config => config)` - Called before loading payment methods.
```
